### PR TITLE
migrate legacy API

### PR DESCRIPTION
**Pull request by wansir**
Signed-off-by: hongming <talonwan@yunify.com>

### CODE REVIEW

File: pkg/models/resources/v1alpha3/application/applications.go [hunk 1]
```diff
@@ -76,3 +82,13 @@ func (d *applicationsGetter) filter(object runtime.Object, filter query.Filter)
 
 	return v1alpha3.DefaultObjectMetaFilter(application.ObjectMeta, filter)
 }
+
+func lastUpdateTime(deployment *appv1beta1.Application) time.Time {
```
Comment (zryfish) [inline pkg/models/resources/v1alpha3/application/applications.go:86]: deployment -> application

File: pkg/kapis/iam/v1alpha2/handler.go [hunk 2]
```diff
@@ -43,108 +37,198 @@ func (h *iamHandler) DescribeUser(req *restful.Request, resp *restful.Response)
...
+					return
+				}
+
+				user = user.DeepCopy()
+
+				if user.Annotations == nil {
+					user.Annotations = make(map[string]string, 0)
+				}
+
+				user.Annotations["iam.kubesphere.io/role"] = roleBinding.RoleRef.Name
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/kapis/iam/v1alpha2/handler.go:155]: just for description only.

File: pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -193,16 +205,26 @@ func (s *APIServer) buildHandlerChain() {
 		handler = filters.WithMultipleClusterDispatcher(handler, clusterDispatcher)
 	}
 
-	excludedPaths := []string{"/oauth/*", "/kapis/config.kubesphere.io/*"}
-	pathAuthorizer, _ := path.NewAuthorizer(excludedPaths)
+	var authorizers authorizer.Authorizer
+
+	switch s.Config.AuthorizationOptions.Mode {
+	case authorizationoptions.AlwaysAllow:
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/apiserver/apiserver.go:210]: AlwaysDenyAuthorizer is useful in unit tests to force an operation to be forbidden. AlwaysAllowAuthorizer is useful in tests and when using kubernetes in an open manner.

File: pkg/apiserver/apiserver.go [hunk 4]
```diff
@@ -118,6 +123,7 @@ func (s *APIServer) PrepareRun() error {
 	s.container.Filter(logRequestAndResponse)
 	s.container.Router(restful.CurlyRouter{})
 	s.container.RecoverHandler(func(panicReason interface{}, httpWriter http.ResponseWriter) {
+		klog.Error(panicReason)
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/apiserver/apiserver.go]: fixed

File: pkg/apiserver/query/types_test.go [hunk 5]
```diff
@@ -21,19 +21,10 @@ func TestParseQueryParameter(t *testing.T) {
...
-					{
-						FieldLabel,
-						Value("app.kubernetes.io/name:book"),
-					},
-					{
-						FieldStatus,
-						Value("Running"),
-					},
+				Filters: map[Field]Value{
+					FieldLabel:  Value("app.kubernetes.io/name:book"),
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/apiserver/query/types_test.go]: fixed

File: pkg/models/resources/v1alpha3/application/applications.go (hunk 1 above)
Comment (wansir) [reply pkg/models/resources/v1alpha3/application/applications.go]: fixed

File: pkg/apiserver/authorization/authorizerfactory/opa.go [hunk 6]
```diff
@@ -33,125 +34,140 @@ type opaAuthorizer struct {
...
-		return authorizer.DecisionDeny, "", err
+		return authorizer.DecisionNoOpinion, "", nil
 	}
 
-	// it's not in workspace resource, permission denied
-	if attr.GetNamespace() == "" {
-		return authorizer.DecisionDeny, permissionUndefined, nil
+	clusterRole, err := o.am.GetClusterRoleOfUser(attr.GetUser().GetName(), attr.GetCluster())
+
+	if errors.IsNotFound(err) {
```
Comment (zryfish) [inline pkg/apiserver/authorization/authorizerfactory/opa.go:103]: check err != nil first

File: pkg/apiserver/authorization/authorizerfactory/opa.go [hunk 7]
```diff
@@ -33,125 +34,140 @@ type opaAuthorizer struct {
...
+
+	// The policy decision is contained in the results returned by the Eval() call. You can inspect the decision and handle it accordingly.
+	results, err := query.Eval(context.Background(), rego.EvalInput(a))
+
+	if err != nil {
+		klog.Errorf("syntax error:%s,refer: %s+v", err, role)
+		return authorizer.DecisionNoOpinion, "", err
+	}
+
+	if len(results) > 0 && results[0].Expressions[0].Value == true {
```
Comment (zryfish) [inline pkg/apiserver/authorization/authorizerfactory/opa.go:166]: Don't we need to check `len(results[0].Expressions) > 0` first?

File: pkg/apiserver/config/config.go [hunk 8]
```diff
@@ -59,19 +60,20 @@ const (
...
-	AuthenticationOptions *authoptions.AuthenticationOptions `json:"-" yaml:"authentication,omitempty" mapstructure:"authentication"`
-	MultiClusterOptions   *multicluster.Options              `json:"multicluster,omitempty" yaml:"multicluster,omitempty" mapstructure:"multicluster"`
+	DevopsOptions         *jenkins.Options                           `json:"devops,omitempty" yaml:"devops,omitempty" mapstructure:"devops"`
+	SonarQubeOptions      *sonarqube.Options                         `json:"sonarqube,omitempty" yaml:"sonarQube,omitempty" mapstructure:"sonarqube"`
+	KubernetesOptions     *k8s.KubernetesOptions                     `json:"kubernetes,omitempty" yaml:"kubernetes,omitempty" mapstructure:"kubernetes"`
+	ServiceMeshOptions    *servicemesh.Options                       `json:"servicemesh,omitempty" yaml:"servicemesh,omitempty" mapstructure:"servicemesh"`
+	NetworkOptions        *network.Options                           `json:"network,omitempty" yaml:"network,omitempty" mapstructure:"network"`
+	LdapOptions           *ldap.Options                              `json:"ldap,omitempty" yaml:"ldap,omitempty" mapstructure:"ldap"`
```
Comment (zryfish) [inline pkg/apiserver/config/config.go:68]: Unnecessary to expose `ldap/redis/s3/authentication/authorization` in `/configz`, use slash to skip

File: pkg/kapis/iam/v1alpha2/register.go [hunk 9]
```diff
@@ -29,97 +29,70 @@ import (
...
 func AddToContainer(container *restful.Container, im im.IdentityManagementInterface, am am.AccessManagementInterface, options *authoptions.AuthenticationOptions) error {
 	ws := runtime.NewWebService(GroupVersion)
 
 	handler := newIAMHandler(im, am, options)
 
+	// global resource
+	ws.Route(ws.GET("/users").
+		To(handler.ListUsers).
+		Doc("List all users.").
+		Param(ws.PathParameter("user", "username")).
```
Comment (zryfish) [inline pkg/kapis/iam/v1alpha2/register.go:50]: no path parameter here

File: pkg/kapis/tenant/v1alpha2/handler.go [hunk 10]
```diff
@@ -15,15 +16,16 @@ type tenantHandler struct {
 	tenant tenant.Interface
 }
 
-func newTenantHandler(k8sClient k8s.Client, factory informers.InformerFactory) *tenantHandler {
+func newTenantHandler(_ k8s.Client, factory informers.InformerFactory) *tenantHandler {
```
Comment (zryfish) [inline pkg/kapis/tenant/v1alpha2/handler.go:19]: If parameter k8sClient is not used anymore, why not remove it from the function signature?

File: pkg/models/iam/am/am.go [hunk 11]
```diff
@@ -19,116 +19,357 @@ package am
...
+func (am *amOperator) GetRoleReferenceRules(roleRef rbacv1.RoleRef, bindingNamespace string) ([]rbacv1.PolicyRule, error) {
+	switch roleRef.Kind {
+	case "Role":
+		role, err := am.k8sinformer.Rbac().V1().Roles().Lister().Roles(bindingNamespace).Get(roleRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		return role.Rules, nil
+
+	case "ClusterRole":
```
Comment (zryfish) [inline pkg/models/iam/am/am.go:354]: Hard cord`Role` and `ClusterRole` here is not a good practice

File: pkg/models/iam/im/ldap_operator.go [hunk 12]
```diff
@@ -78,3 +79,7 @@ func (im *ldapOperator) CreateUser(user *iamv1alpha2.User) (*iamv1alpha2.User, e
 
 	return user, nil
 }
+
+func (im *ldapOperator) ListUsers(query *query.Query) (*api.ListResult, error) {
```
Comment (zryfish) [inline pkg/models/iam/im/ldap_operator.go:83]: What's this for

File: pkg/apiserver/authorization/authorizerfactory/opa.go (hunk 6 above)
Comment (wansir) [reply pkg/apiserver/authorization/authorizerfactory/opa.go:103]: It's unnecessary, if err is nil, always false.

File: pkg/apiserver/authorization/authorizerfactory/opa.go (hunk 7 above)
Comment (wansir) [reply pkg/apiserver/authorization/authorizerfactory/opa.go:166]: It's not needed, already checked syntax error.

File: pkg/apiserver/config/config.go (hunk 8 above)
Comment (wansir) [reply pkg/apiserver/config/config.go]: done

File: pkg/models/iam/im/ldap_operator.go (hunk 12 above)
Comment (wansir) [reply pkg/models/iam/im/ldap_operator.go:84]: Not used yet, might be removed.

File: pkg/kapis/tenant/v1alpha2/handler.go (hunk 10 above)
Comment (wansir) [reply pkg/kapis/tenant/v1alpha2/handler.go]: done

File: pkg/models/iam/am/am.go (hunk 11 above)
Comment (wansir) [reply pkg/models/iam/am/am.go]: done

**Pull request merged**